feat: add view count tracking feature for blog post

CynthiaWahome opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

​

## Description

This PR implements a blog view count feature that tracks how many times each blog post has been viewed. The counter increments each time a blog post is retrieved through the API endpoint.

​

## Related Issue (Link to issue ticket)
Closes #1000
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

​

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

​Adding view counts to blog posts provides valuable analytics for content creators and administrators. This information helps identify popular content and improve the overall user experience by highlighting trending articles.


## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

- Added unit tests that verify view count increments correctly with each blog view
- Manually tested with Postman to confirm the feature works in a real API environment
- Verified that the counter increases across multiple requests to the same blog post
- Ensured backward compatibility with existing endpoints
​

## Screenshots:

<img width="689" alt="image" src="https://github.com/user-attachments/assets/8448ef67-0e9a-4768-864a-b1b99a5cd15d" />

​

<img width="689" alt="image" src="https://github.com/user-attachments/assets/79172762-9ccb-4069-ba13-f6e9893f4177" />


## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
      ​

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

CynthiaWahome committed:
feat: Add view count tracking to Blog model

Summary:
- Added Integer import from SQLAlchemy at the top of the file
- Added 'views' column to Blog model with Integer type
- Set default value to 0 and not nullable via server_default
- Added explanatory comment for the new column
- Ensured compatibility with existing database schema

CynthiaWahome committed:
feat: Implement blog view count tracking service

Summary:
- Added increment_view_count method to BlogService
- Refactored code to follow DRY principles by abstracting view count logic
- Added support for both dictionary and object-style blogs for testing
- Implemented proper error handling with try/except and database transaction management
- Added comprehensive method documentation with docstrings
- Commented out redundant code to improve maintainability

CynthiaWahome committed:
feat: Include view count in blog response schemas

Summary:
- Created BlogBaseResponse as a reusable base class following DRY principles
- Updated BlogResponse to inherit from base class to avoid duplication
- Added proper validation and type hints for view count field
- Refactored schema architecture to improve maintainability
- Maintained backward compatibility while improving code structure

CynthiaWahome committed:
feat: Integrate blog view counter in API endpoint

Summary:
- Updated blog fetch endpoint to call increment_view_count
- Added view count tracking functionality to blog retrieval flow
- Ensured views are incremented only once per request
- Maintained backward compatibility with existing API consumers

CynthiaWahome committed:
test: Add comprehensive testing for blog view counter

Summary:
- Created test_blog_view_count_increments test function to validate the feature
- Implemented mock blog simulation to mimic database behavior without database dependency
- Set up stateful testing to verify views increment across multiple requests
- Adapted testing approach to handle both dictionary-based mocks and object models
- Configured mock database session to return updated view count values between calls
- Verified all assertions pass with the correct incrementation logic

CynthiaWahome committed:
test: add dedicated test for blog view count incrementation

Summary:
- Added test_blog_view_count_increments function to verify view count behavior
- Implemented stateful testing approach to simulate multiple blog views
- Added assertions to verify counter increments correctly with each view
- Updated mocks between requests to simulate database persistence
- Ensured compatibility with the service implementation

Signed-off-by: Cynthia <160864573+CynthiaWahome@users.noreply.github.com>

CynthiaWahome committed:
test: add views field to mock blog object

Signed-off-by: Cynthia <160864573+CynthiaWahome@users.noreply.github.com>

samuelogboye commented on api/v1/routes/blog.py:
```diff
@@ -77,7 +77,8 @@ def get_blog_by_id(id: str, db: Session = Depends(get_db)):
 
-    blog_post = blog_service.fetch(id)
+    # Fetch blog and increment view count
+    blog_post = blog_service.increment_view_count(id)
```
making this increment_view_count makes it seem as if the only thing the method does is to increase view count.

CynthiaWahome replied:
@samuelogboye Thanks for the feedback. 
I'd like to make sure I understand your concern correctly.

You mentioned that the name `increment_view_count` makes it seem like the method only increases the view count, when it actually:
1. Fetches the blog post
2. Increments the view count 
3. Returns the updated blog

Is your suggestion that I should rename this method to better reflect all of its functionality? If so, I'm happy to update it with a more descriptive name.

CynthiaWahome replied:
@samuelogboye  I've addressed your feedback about the method name by renaming it from `increment_view_count` to `fetch_and_increment_view` to better reflect that it both fetches the blog and increments its view count. 

Let me know if this addresses your concern, thanks.

samuelogboye commented:
LGTM!

CynthiaWahome committed:
Merge branch 'dev' into feature/blog-view-count

Takinnuoye5 reviewed (changes requested):
you need to fix this : ensure that the increment_view_count method handles edge cases, such as when the blog ID does not exist like Add custom error messages (e.g., "Blog not found" instead of generic Exception).

also you need to Call self.db.refresh(blog) before commit to avoid unnecessary database writes if refresh fails.

CynthiaWahome committed:
refactor: Enhance blog view counter with improved error handling

Summary:
- Renamed method to fetch_and_increment_view for better clarity
- Added edge case handling for non-existent blog IDs
- Implemented custom error messages instead of generic exceptions
- Reordered database operations to refresh before commit
- Added proper error logging for troubleshooting
- Updated route references to use new method name

CynthiaWahome commented:
@Takinnuoye5  I've addressed your feedback:

1. Added custom error handling for non-existent blogs with specific error messages.
2. Reordered operations to refresh **BEFORE** commit to ensure data integrity.

Let me know if this addresses your concerns, thanks.

CynthiaWahome committed:
Merge branch 'dev' into feature/blog-view-count

CynthiaWahome commented:
Thank you @Takinnuoye5 @samuelogboye 
Appreciate it.
